bgpd/bgp_route.c

```diff
@@ -10936,16 +10981,71 @@ DEFPY (show_ip_bgp_json,
 		if (argv_find(argv, argc, "exact-match", &idx))
 			exact_match = 1;
 
+		if (!community)
+			sh_type = bgp_show_type_community_all;
+	}
+
+	if (!all) {
+		/* show bgp: AFI_IP6, show ip bgp: AFI_IP */
 		if (community)
 			return bgp_show_community(vty, bgp, community,
-						  exact_match, afi, safi, uj);
+						  exact_match, afi, safi,
+						  sh_flags);
 		else
-			return (bgp_show(vty, bgp, afi, safi,
-					 bgp_show_type_community_all, NULL, uj,
-					 wide));
-	}
+			return (bgp_show(vty, bgp, afi, safi, sh_type, NULL,
+					 sh_flags));
+	} else {
+		/* show <ip> bgp ipv4 all: AFI_IP, show <ip> bgp ipv6 all:
+		 * AFI_IP6 */
+		if (CHECK_FLAG(sh_flags, BGP_SHOW_OPT_AFI_IP)
+		    || CHECK_FLAG(sh_flags, BGP_SHOW_OPT_AFI_IP6)) {
+			afi = CHECK_FLAG(sh_flags, BGP_SHOW_OPT_AFI_IP)
+				      ? AFI_IP
+				      : AFI_IP6;
+			FOREACH_SAFI(safi)
+			{
+				if (strmatch(get_afi_safi_str(afi, safi, true),
+					     "Unknown"))
+					continue;
+
+				if (!bgp_afi_safi_peer_exists(bgp, afi, safi))
+					continue;
 
-	return bgp_show(vty, bgp, afi, safi, sh_type, NULL, uj, wide);
+				vty_out(vty, "\nFor address family: %s\n\n",
```

Comment:
is this vty output needed? What if I wanna print JSON? This will be printed as well together with JSON?

Reply:
router# show ip bgp ipv6 all json

For address family: IPv6 Unicast

{
 "vrfId": 0,
 "vrfName": "default",
 "tableVersion": 3,
 "routerId": "1.1.1.1",
 "defaultLocPrf": 100,
 "localAS": 1,
 "routes": { "2001:db8::1/128": [
  {
    "valid":true,
    "bestpath":true,
    "pathFrom":"external",
    "prefix":"2001:db8::1",
    "prefixLen":128,
    "network":"2001:db8::1\/128",
    "metric":0,
    "weight":32768,
    "peerId":"(unspec)",
    "path":"",
    "origin":"incomplete",
    "nexthops":[
      {
        "ip":"::",
        "hostname":"router",
        "afi":"ipv6",
        "scope":"global",
        "used":true
      }
    ]
  }
],"2001:db8::2/128": [
  {
    "valid":true,
    "bestpath":true,
    "pathFrom":"internal",
    "prefix":"2001:db8::2",
    "prefixLen":128,
    "network":"2001:db8::2\/128",
    "metric":0,
    "locPrf":100,
    "weight":0,
    "peerId":"192.168.56.152",
    "path":"",
    "origin":"incomplete",
    "nexthops":[
      {
        "ip":"fe80::a00:27ff:fefc:2aa",
        "hostname":"router",
        "afi":"ipv6",
        "scope":"global",
        "used":true
      },
      {
        "ip":"fe80::a00:27ff:fefc:2aa",
        "hostname":"router",
        "afi":"ipv6",
        "scope":"link-local"
      }
    ]
  }
],"2001:db8:85a3::8a2e:370:7334/128": [
  {
    "valid":true,
    "bestpath":true,
    "pathFrom":"external",
    "prefix":"2001:db8:85a3::8a2e:370:7334",
    "prefixLen":128,
    "network":"2001:db8:85a3::8a2e:370:7334\/128",
    "metric":0,
    "weight":32768,
    "peerId":"(unspec)",
    "path":"",
    "origin":"incomplete",
    "nexthops":[
      {
        "ip":"::",
        "hostname":"router",
        "afi":"ipv6",
        "scope":"global",
        "used":true
      }
    ]
  }
] }  } 
router#

Reply:
so, yes, we must remove vty_out for json stuff.

Reply:
ok will skip the below line for json.  
vty_out(vty, "\nFor address family: %s\n\n", get_afi_safi_str(afi, safi, false));